lib/AST/LookupVisibleDecls.cpp

```diff
@@ -473,19 +474,96 @@ lookupVisibleMemberDeclsImpl(Type BaseTy, VisibleDeclConsumer &Consumer,
                              GenericSignatureBuilder *GSB,
                              VisitedSet &Visited);
 
-static void lookupVisibleProtocolMemberDecls(
-    Type BaseTy, ProtocolType *PT, VisibleDeclConsumer &Consumer,
-    const DeclContext *CurrDC, LookupState LS, DeclVisibilityKind Reason,
-                                             LazyResolver *TypeResolver, GenericSignatureBuilder *GSB,
-    VisitedSet &Visited) {
+class RestateFilteringConsumer : public VisibleDeclConsumer {
+  VisibleDeclConsumer &parentConsumer;
+  LazyResolver *resolver;
+
+  llvm::DenseSet<DeclName> foundVars;
+  llvm::DenseSet<std::pair<DeclName, CanType>> foundFuncs;
+
+public:
+  RestateFilteringConsumer(VisibleDeclConsumer &parentConsumer,
+                           Type baseTy, const DeclContext *DC,
+                           LazyResolver *resolver)
+  : parentConsumer(parentConsumer), resolver(resolver) {
+    assert(DC && baseTy && !baseTy->hasLValueType());
+  }
+
+  void foundDecl(ValueDecl *VD, DeclVisibilityKind Reason) override {
+    assert(VD);
+    // If this isn't a protocol context, hand over the decl
+    // to the parent consumer.
+    if (!isa<ProtocolDecl>(VD->getDeclContext())) {
+      parentConsumer.foundDecl(VD, Reason);
+      return;
+    }
+    if (resolver)
+      resolver->resolveDeclSignature(VD);
+
+    if (!VD->hasInterfaceType()) {
+      parentConsumer.foundDecl(VD, Reason);
+      return;
+    }
+    auto AFT = VD->getInterfaceType()->getAs<AnyFunctionType>();
+    if (!AFT) {
+      if (foundVars.insert(VD->getFullName()).second)
+        parentConsumer.foundDecl(VD, Reason);
+      return;
+    }
+    auto type = stripSelfRequirementsIfNeeded(VD, AFT);
+    if (foundFuncs.insert({VD->getFullName(), type}).second) {
+      parentConsumer.foundDecl(VD, Reason);
+    }
+  }
+private:
+  CanType stripSelfRequirementsIfNeeded(ValueDecl *VD, AnyFunctionType *AFT) {
+    // Preserve the generic signature if this is a subscript, which are uncurried,
+    // or if we have genenic params other than Self. Otherwise, strip if off
+    // and use the resultType of the curried function type.
+    // When preserving the generic signature, we remove the requirements
+    // from Self to make they don't prevent us from recognizing restatements.
+
+    // This can't be null since we are dealing with method or
```

Comment:
In that case, the function should take a GenericFunctionType instead of an AnyFunctionType, then this comment becomes unnecessary